### PR TITLE
DIT-1025 - sample changes for referred case for sample

### PR DESCRIPTION
**Pull request by SteveSugden**

### CODE REVIEW

File: app/uk/gov/hmrc/tariffclassificationfrontend/views/partials/sample/sample_status_events_table.scala.html [hunk 1]
```diff
@@ -44,7 +44,16 @@ <h3 class="heading-medium" id="sample-status-events-heading">Sample activity</h3
                             val body = event.details.asInstanceOf[SampleStatusChange]
                             progressive_disclosure(
                                 id = s"sample-status-events-row-$index",
-                                summary = SampleStatus.format(body.to),
+                                summary = s"Sample status changed from ${SampleStatus.format(body.from, false)} to ${SampleStatus.format(body.to, false)}",
+                                body = body.comment.map(Html(_))
```
Comment (aaroncarroll82) [inline app/uk/gov/hmrc/tariffclassificationfrontend/views/partials/sample/sample_status_events_table.scala.html]: It didn't support it at the time of you putting this in - but progressive disclosure can now support tabindexes, so we should probably set them here to allow for expanding these elements using keyboard nav.

Comment (SteveSugden) [reply app/uk/gov/hmrc/tariffclassificationfrontend/views/partials/sample/sample_status_events_table.scala.html]: Good spot

File: conf/app.routes [hunk 2]
```diff
@@ -87,6 +87,9 @@ POST        /cases/:reference/change-appeal/:appealId           @uk.gov.hmrc.tar
 GET         /cases/:reference/sample/status                     @uk.gov.hmrc.tariffclassificationfrontend.controllers.SampleController.chooseStatus(reference: String)
 POST        /cases/:reference/sample/status                     @uk.gov.hmrc.tariffclassificationfrontend.controllers.SampleController.updateStatus(reference: String)
 
+GET         /cases/:reference/sample/returnStatus               @uk.gov.hmrc.tariffclassificationfrontend.controllers.SampleReturnController.chooseStatus(reference: String)
+POST        /cases/:reference/sample/returnStatus               @uk.gov.hmrc.tariffclassificationfrontend.controllers.SampleReturnController.updateStatus(reference: String)
```
Comment (aaroncarroll82) [inline conf/app.routes]: Should these endpoints have camelcase in? Looks like we typically hyphenate when multi word. Thought there was possibly some kind of accessibility concern around camel cased URLs?

File: app/uk/gov/hmrc/tariffclassificationfrontend/views/partials/sample/sample_status_events_table.scala.html (hunk 1 above)
Comment (SteveSugden) [reply app/uk/gov/hmrc/tariffclassificationfrontend/views/partials/sample/sample_status_events_table.scala.html]: Fixed.   But the progressive bit doesn't appear to be used 'cos there is no 'comment'.  But done anyway

File: conf/app.routes (hunk 2 above)
Comment (SteveSugden) [reply conf/app.routes]: Another good spot.  Fixed

Comment (SteveSugden) [reply conf/app.routes]: in my defence, I got confused with the camelCase variable names (not same thing I know)

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/SampleReturn.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  type SampleReturn = Value
+
+  val YES, NO, TO_BE_CONFIRMED = Value
+
+  def format(status: Option[SampleReturn]): String = {
+    status match {
+      case Some(YES) => "Yes"
+      case Some(NO) => "No"
+      case Some(TO_BE_CONFIRMED) => "To be confirmed"
+      case _ => "Unknown"
```
Comment (aaroncarroll82) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/SampleReturn.scala]: Was just wondering with this - should case of None just be 'To be confirmed'? Still makes sense, but possibly looks  bit more pleasant than Unknown

Comment (SteveSugden) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/SampleReturn.scala]: Should never happen.  If the status is None then it shouldn't be rendered.

Comment (pllaurado) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/SampleReturn.scala]: should never happen = thrown exception ?

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/SampleReturn.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package uk.gov.hmrc.tariffclassificationfrontend.models
+
+object SampleReturn extends Enumeration {
+  type SampleReturn = Value
+
+  val YES, NO, TO_BE_CONFIRMED = Value
+
+  def format(status: Option[SampleReturn]): String = {
```
Comment (pllaurado) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/SampleReturn.scala]: I prefer to use Options just to return or store when something can be empty, but create a function with option param ... is just an opinion

File: app/uk/gov/hmrc/tariffclassificationfrontend/controllers/SampleReturnController.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                                       override val messagesApi: MessagesApi,
+                                       override implicit val config: AppConfig) extends StatusChangeAction[Option[SampleReturn]] {
+
+  override protected val requiredPermission: Permission.Value = Permission.EDIT_SAMPLE
+
+  override protected def redirect: String => Call = routes.CaseController.trader
+
+  override protected def isValidCase(c: Case)(implicit request: AuthenticatedRequest[_]): Boolean = {
+    true //No constraints on when the case is valid
+  }
```
Comment (ed0906) [inline app/uk/gov/hmrc/tariffclassificationfrontend/controllers/SampleReturnController.scala]: Minor but the braces arent required

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Sample.scala [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package uk.gov.hmrc.tariffclassificationfrontend.models
+
+import uk.gov.hmrc.tariffclassificationfrontend.models.SampleReturn.SampleReturn
+import uk.gov.hmrc.tariffclassificationfrontend.models.SampleStatus.SampleStatus
+
+case class Sample
+(
+  status: Option[SampleStatus] = None,
```
Comment (ed0906) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/Sample.scala]: Option? how come?  Isnt `Sample` Optional on the Case level?

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Sample.scala]: I take it back, it  isnt

**Pull request merged**